src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/segmentation/KernelSegmenter.java

```diff
@@ -0,0 +1,526 @@
+                    "using all data points to calculate kernel matrix.");
+        }
+
+        //subsample data with replacement
+        final int numSubsample = Math.min(kernelApproximationDimension, data.size());
+        logger.info(String.format("Subsampling %d points from data to find kernel approximation...", numSubsample));
+        final List<T> dataSubsample = numSubsample == data.size()
+                ? data
+                : IntStream.range(0, numSubsample).boxed().map(i -> data.get(rng.nextInt(data.size()))).collect(Collectors.toList());
+
+        //calculate (symmetric) kernel matrix of subsampled data
+        logger.info(String.format("Calculating kernel matrix of subsampled data (%d x %d)...", numSubsample, numSubsample));
+        final RealMatrix subKernelMatrix = new Array2DRowRealMatrix(numSubsample, numSubsample);
+        for (int i = 0; i < numSubsample; i++) {
+            for (int j = 0; j < i; j++) {
+                final double value = kernel.apply(dataSubsample.get(i), dataSubsample.get(j));
+                subKernelMatrix.setEntry(i, j, value);
+                subKernelMatrix.setEntry(j, i, value);
+            }
+            subKernelMatrix.setEntry(i, i, kernel.apply(dataSubsample.get(i), dataSubsample.get(i)));
+        }
+
+        //perform SVD of kernel matrix of subsampled data
+        logger.info(String.format("Performing SVD of kernel matrix of subsampled data (%d x %d)...", numSubsample, numSubsample));
+        final SingularValueDecomposition svd = new SingularValueDecomposition(subKernelMatrix);
+
+        //calculate reduced observation matrix
+        logger.info(String.format("Calculating reduced observation matrix (%d x %d)...", data.size(), numSubsample));
+        final RealMatrix reducedObservationMatrix = new Array2DRowRealMatrix(data.size(), numSubsample);
+        reducedObservationMatrix.walkInOptimizedOrder(new DefaultRealMatrixChangingVisitor() {
+            final double[] invSqrtSingularValues = Arrays.stream(svd.getSingularValues()).map(Math::sqrt).map(x -> 1. / (x + EPSILON)).toArray();
+            @Override
+            public double visit(int i, int j, double value) {
+                double sum = 0.;
+                for (int k = 0; k < numSubsample; k++) {
+                    sum += kernel.apply(data.get(i), dataSubsample.get(k)) * svd.getU().getEntry(k, j) * invSqrtSingularValues[j];
+                }
+                return sum;
+            }
+        });
+        return reducedObservationMatrix;
+    }
+
+    //for N x p matrix Z_ij, returns the N-dimensional vector sum(Z_ij * Z_ij, j = 0,..., p - 1),
+    //which are the diagonal elements K_ii of the approximate kernel matrix
+    private static double[] calculateKernelApproximationDiagonal(final RealMatrix reducedObservationMatrix) {
+        return IntStream.range(0, reducedObservationMatrix.getRowDimension()).boxed()
```

Comment:
It's lighter and shorter to use `IndexRange`:
```
return new IndexRange(0, reducedObservationMatrix.getRowDimension())
   .mapToDouble(i -> Arrays.stream(reducedObservationMatrix.getRow(i)).map(z -> z * z).sum());
```
You could go a step further with
```
return new IndexRange(0, reducedObservationMatrix.getRowDimension())
   .mapToDouble(i -> MathUtils.square(reducedObservationMatrix.getRowVector(i).getNorm()));
```

Reply:
Done.